src/Chapter3.hs

```diff
@@ -560,22 +679,34 @@ introducing extra newtypes.
 🕯 HINT: if you complete this task properly, you don't need to change the
     implementation of the "hitPlayer" function at all!
 -}
+
+newtype Attack    = Attack    Int deriving (Show)
+newtype Armor     = Armor     Int deriving (Show)
+newtype Health    = Health    Int deriving (Show)
+newtype Dexterity = Dexterity Int deriving (Show)
+newtype Strength  = Strength  Int deriving (Show)
+newtype Damage    = Damage    Int deriving (Show)
+newtype Defense   = Defense   Int deriving (Show)
+
 data Player = Player
-    { playerHealth    :: Int
-    , playerArmor     :: Int
-    , playerAttack    :: Int
-    , playerDexterity :: Int
-    , playerStrength  :: Int
-    }
+    {playerHealth    :: Health
+    ,playerArmor     :: Armor
+    ,playerAttack    :: Attack
+    ,playerDexterity :: Dexterity
+    ,playerStrength  :: Strength
+    } deriving (Show)
 
-calculatePlayerDamage :: Int -> Int -> Int
-calculatePlayerDamage attack strength = attack + strength
+calculatePlayerDamage :: Attack -> Strength -> Damage
+calculatePlayerDamage (Attack attack') (Strength strength)
+    = Damage (attack' + strength)
 
-calculatePlayerDefense :: Int -> Int -> Int
-calculatePlayerDefense armor dexterity = armor * dexterity
+calculatePlayerDefense :: Armor -> Dexterity -> Defense
+calculatePlayerDefense (Armor armor) (Dexterity dexterity)
+    = Defense (armor * dexterity)
 
-calculatePlayerHit :: Int -> Int -> Int -> Int
-calculatePlayerHit damage defense health = health + defense - damage
+calculatePlayerHit :: Damage -> Defense -> Health -> Health
+calculatePlayerHit (Damage damage) (Defense defense) (Health health)
+    = Health (health + defense - damage)
```

Comment:
Nice work on newtypes! :medal_sports: Newtypes are one very important feature in Haskell :+1:
You can read more about the newtype pattern in our blog post: https://kowainik.github.io/posts/haskell-mini-patterns#newtype

Reply:
I already started looking at the website. Amazing job !